### PR TITLE
Progress bar for inference

### PR DESCRIPTION
**Pull request by b5y**
#### Summary:
Progress bar for inference. Supports Rejection and BOLFI samplers (in other samplers it was impossible to calculate the number of simulations).

---
This change is [Reviewable](https://reviewable.io/reviews/elfi-dev/elfi/281)

### CODE REVIEW

**Comment by vuolleko** [comment]
Have you considered using the number of samples instead of the number of simulations?

**Comment by b5y** [comment]
Yes, I have. It crashes with BOLFI with the number of samples. Other samplers work perfectly fine with the number of samples. 

The main idea was about covering all samplers and showing the process of simulation. That's why it is quite complicated now. If it is okay to keep progress bar for BOLFI with the number of simulations and other samplers with the number of samples, then I can rewrite it.

**Comment by vuolleko** [comment]
I suppose BO and BOLFI crash because calling `fit` or `infer` don't *sample*. So with those the number of simulations is certainly appropriate. Sampling is done via MCMC; feel free to add a progress bar therein (can be a separate PR as the system is currently quite independent of ELFI).

But most other samplers should be easier to do with respect to the number of samples instead of the number of simulations.

File: elfi/visualization/progress_bar.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+from math import ceil
+
+from IPython.display import display
+from ipywidgets import HTML, IntProgress, VBox
+
+
+class ProgressBar(object):
+    """Progress bar for showing the inference process."""
+
+    def __init__(self, batch_size, n_samples, sampler, quantile=None, n_sim=None):
```
Comment (vuolleko) [inline elfi/visualization/progress_bar.py]: Making the progress bar depend on all this info doesn't make it very general. 

Testing with a `print` in `ParameterInference.infer` should give you an idea for a rather general solution (inspired by `ParameterInference.finished`):

```
while not self.finished:
    print('Progress: ', self.state['n_batches'], self._objective_n_batches)
    self.iterate()
```

If you always update the ProgressBar with both ` self.state['n_batches']` and  `self._objective_n_batches`, it should work properly in all cases. With threshold-based methods users may sometimes observe "negative" progress, but I guess it's fine.

File: elfi/visualization/progress_bar.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        self.bar = IntProgress(min=0, max=self.n_iter)
+        self.label = HTML()
+        self.box = VBox(children=[self.label, self.bar])
+        display(self.box)
+
+    def update(self):
+        """Update the step in progress bar."""
+        self.bar.value += self.step
+        self.label.value = '{name}: {index} / {size}'.format(
+            name="Inference progress bar (number of simulations)",
```
Comment (vuolleko) [inline elfi/visualization/progress_bar.py]: Too verbose. Maybe just "Progress"?

File: elfi/visualization/progress_bar.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        elif sampler == 'bolfi':
+            self.n_iter = n_sim
+            self.step = ceil(n_sim / 100)
+        else:
+            self.n_iter = n_sim if n_sim else n_samples / batch_size
+            self.step = self.n_iter / 100
+        self.bar = IntProgress(min=0, max=self.n_iter)
+        self.label = HTML()
+        self.box = VBox(children=[self.label, self.bar])
+        display(self.box)
```
Comment (vuolleko) [inline elfi/visualization/progress_bar.py]: Would be nice to get textual status if `display` is unsupported.

File: elfi/methods/parameter_inference.py [hunk 4]
```diff
@@ -464,7 +468,13 @@ def set_objective(self, n_samples, threshold=None, quantile=None, n_sim=None):
         else:
             n_batches = self.max_parallel_batches
 
-        self.objective = dict(n_samples=n_samples, threshold=threshold, n_batches=n_batches)
+        self.objective = dict(n_samples=n_samples,
```
Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: Why this change?

**Comment by vuolleko** [comment]
This feature was requested in #254
